feat: allowing routes to be filtered and adding basic example (#715)
VeryGoodOpenSource/dart_frog: packages/dart_frog_auth/README.md

> In the case of `null` be returned (unauthenticated), the middleware will automatically send an unauthorized `401` in the response.

Reviewer: I have a few changes I'd like to make to this README so I'll wait until after this is merged to do so

> ### Filtering routes

Reviewer: ```suggestion
### Filtering Routes
```

In many instances, developers will want to apply authentication to some routes, while not to others.

> One of those can be described by looking at implementing a basic RESTful CRUD API. In order to make
> such an API that allows consumers to create, update, delete and get user information, the following list
> of routes will need to be created:

Reviewer: ```suggestion
One of those can be described by looking at implementing a basic RESTful CRUD API. In order to make
such an API that allows consumers to create, update, delete, and get user information, the following list
of routes will need to be created:
```

>  - `POST /users`: Creates an user
>  - `PATCH /users/[id]`: Updates the user with the given id.
>  - `DELETE /users/[id]`: Deletes the user with the given id.
>  - `GET /users/[id]`: Returns the user with the given id.

Reviewer: ```suggestion
 - `POST /users`: Creates a user
 - `PATCH /users/[id]`: Updates the user with the given id.
 - `DELETE /users/[id]`: Deletes the user with the given id.
 - `GET /users/[id]`: Returns the user with the given id.
```

> Those endpoints can be translated in the following structure in a Dart Frog backend:

Reviewer: ```suggestion
Those endpoints can be translated to the following structure in a Dart Frog backend:
```

only an authenticated user would be allowed to change those information.

To accomplish that, we need that the middleware apply authentication to all routes, except the `POST`.

Such behavior is possible with the use of `applyToRoute` optional predicate:

```dart
Handler middleware(Handler handler) {
  final userRepository = UserRepository();

  return handler
      .use(requestLogger())
      .use(provider<UserRepository>((_) => userRepository))
      .use(
        basicAuthentication<User>(
          userFromCredentials: userFromCredentials(userRepository),
          applyToRoute: (RequestContext context) async =>
              context.request.method != HttpMethod.post,
        ),
      );
}
```

In the above example, only routes that are not `POST` will have authentication checked.

### Authentication vs Authorization

Authentication and Authorization are two related, but different concepts that are often confused.

Authentication is about WHO the user is, while authorization is about WHAT a user can do.

These concepts are related since we need to know who the user is, in order to check if they can
perform or not a given operation.

> `dart_frog_auth` tries only to solve the authentication part of the problem, in order to enforce
> authorization, it is up to the developer to implement it manually, or use an authorization issue
> system, like OATH for example.
> 
> In technical terms, a request should return `401` (Unauthorized) when authentication fails and
> `403` (Forbidden) when authorization failed.
> 
> The following snippet shows how authorization can be manually checked in `DELETE /users/[id]` route,
> where only the current logged user is allowed to delete itself:

Reviewer: ```suggestion
`dart_frog_auth` only solves the authentication part of the problem. To enforce
authorization, it is up to the developer to implement it manually, or use an authorization issue
system like OATH, for example.

In technical terms, a request should return `401` (Unauthorized) when authentication fails and
`403` (Forbidden) when authorization failed.

The following snippet shows how authorization can be manually checked in `DELETE /users/[id]` route,
where only the current logged user is allowed to delete itself:
```
